rfc/rfc-0014-binding-injection.md

```diff
@@ -0,0 +1,131 @@
+# RFC-0014: Binding Injection
+
+**Authors:** @scothis
+
+**Status:**
+
+**Pull Request URL:** [#1382](https://github.com/projectriff/riff/pull/1382)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** [RFC 0002 Bindings](./rfc-0002-bindings.md), [RFC 0008 Pod Templates](./rfc-0008-pod-templates.md)
+
+
+## Problem
+
+[RFC 0002 Bindings](./rfc-0002-bindings.md) defines a interface for a resource to expose metadata and secrets to be consumed by another resource. It leverages the [CNB Binding](https://github.com/buildpack/spec/blob/master/extensions/bindings.md) spec for how the service metadata and secrets are exposed inside a container. While a binding can be consumed manually, it's difficult to manage as it requires defining two volumes, and two volume mounts and a environment variable per container.
+
+### Anti-Goals
+
+This RFC has no opinion as to the type of binding being injected, or the consumer of the binding.
+
+## Solution
+
+Knative Bindings (yea, the term is overloaded) define an abstract way for reconcilers to operate on resources by duck typing known elements of the resource without requiring pre-knowledge or full type knowledge at runtime. By leveraging resources that use a PodTemplateSpec, it's possible to inject volumes, volume mounts and environment variables into an existing resource (even if owned and controlled by another resource). With [RFC 0008 Pod Templates](./rfc-0008-pod-templates.md) the riff Deployers and Processor resources implement the PodSpecable duck type.
+
+A duck type for exposing a binding is defined in RFC 0002. While not as wide spread, the Stream resource currently implements this duck type. Combining the Binding and PodSpecable duck types allows for a generic means to inject binding metadata and secrets into a consuming workload. A new CRD can express this injection:
+
+```yaml
+apiVersion: bindings.projectriff.io/v1alpha1
+kind: Injection
+metadata:
+  name: my-binding-injection
+spec:
+  binding:
+    apiVersion: streaming.projectriff.io/v1alpha1
+    kind: Stream
+    name: my-stream
+  containers:
+  - user-container
+  subject:
+    apiVersion: knative.projectriff.io/v1alpha1
+    kind: Deployer
+    name: my-deployer
+status:
+  observedGeneration: 1
+  conditions: []
+```
+
+- `.spec.binding` is an ObjectReference to a resource implementing the binding duck type
```

Comment:
Is there a definition of this duck type somewhere? I'm missing some detail.

Reply:
It's rfc-0002 I guess. If I have comments on that where do they go (since it is "approved" I assume it is basically immutable)?

Reply:
For basic errata we can fx with a PR to the current rfc. To change the semantics or scope, we should create a new rfc that supersedes the rfc in question.

For now, feel free to ask here.

Reply:
OK. It seemed wrong to me in rfc-0002 that the producer and consumer both have to explicitly state the config map and secret refs. In a way, maybe this RFC (0014) is a step towards correcting that - you only need the binding ref to consume it via an `Injection`.

Reply:
It's not a matter of right and wrong. RFC 0002 defined a low level mechanism to support a CNB Bindings in a k8s workload. Defining a high level the user experience wasn't the goal. RFC 0014 is proposing a higher level experience.

Reply:
@dsyer to answer you question about defining duck types, my read of the Knative Bindings [doc here](https://docs.google.com/document/d/1t5WVrj2KQZ2u5s0LvIUtfHnSonBv5Vcv8Gl2k5NXrCQ/edit#heading=h.tur8dkr939w4), is that duck types are the common schema fragments which implementations of Bindings choose to standardize on.